src/analysis/souffle/check_predicate.dl

```diff
@@ -0,0 +1,82 @@
+//-----------------------------------------------------------------------------
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//-----------------------------------------------------------------------------
+#ifndef SRC_ANALYSIS_SOUFFLE_CHECK_PREDICATE_DL_
+#define SRC_ANALYSIS_SOUFFLE_CHECK_PREDICATE_DL_
+
+#include "authorization_logic.dl"
+#include "dataflow_graph.dl"
+#include "tags.dl"
+
+.type Bool <: number
+
+.type CheckPredicate =
```

Comment:
Is my understanding correct that this is more expressive than what we have in C++ representation?

For instance, the C++ representation can only express something like `ap is tag1 or (tag2 and !tag3)`. However, this can represent `ap1 is tag1 or ap2 is tag2 and ap2 is tag3`. That is, we can combine predicates of different access paths. Is that correct?

Reply:
No, this is exactly as expressive as what we have in the C++ representation. Each of the ADT branches correspond directly to a `Predicate` we use within checks on the C++ side.

You might be thinking of `claim`s, where we have retreated from the full expressiveness of Arcs because we're not sure what some of that expressiveness means in the overapproximated world of our analysis (for instance, if we claim that some node has `tag1 or tag2`, what does that mean for downstream nodes.

Reply:
The root `CP_TagPresence` also has an `AccessPath` as an argument. So, "(ap1 is tag1) or (ap2 is tag2)` can be represented as follows: 
```
CP_Or { 
   lhs: CP_TagPresence("ap1", "someone", "tag1"),
   rhs: CP_TagPresence("ap2", "someone", "tag2")
}
```

no? On the other hand, the C++ `TagPresence` only has a tag: 
 https://github.com/google-research/raksha/blob/ba14ca7d5a4aebbe31405bc39164dbd6079221e1/src/ir/predicate.h#L232

Reply:
Oh, I guess you are right. It is more expressive than the original in being able to speak about different `AccessPath`s in the same predicate. I hadn't noticed that.

Another interesting note: because it explicitly lists the `owner` predicate, it can talk about different `owner`s within the same expression tree as well. `owner`s cannot be explicitly listed in the current C++ representation. Currently, the C++ representation translates to a Datalog rule which accepts any owner of the given `tag`; I was thinking of adding a `TagPresenceDynamicOwner` to capture this behavior. I think it is valuable to have a way to require a particular owner as well, hence why I'm thinking of adding an alternative version of `TagPresence`.

Reply:
Sounds good. This is a good change in terms of expressiveness as long as it doesn't blow up the performance. :) Based on your other comment, it seems like we would be OK.